[#148570883] Enable Postgres extensions

henrytk opened the pull request:
## What

https://www.pivotaltracker.com/n/projects/1275640/stories/148570883

See commit messages for full details.

## How to review

* code review
* see the README for running unit tests.
* check our build CI for the integration tests of this code
* if you want to deploy a bosh release has been published and can be deployed using this branch of paas-cf: https://github.com/alphagov/paas-cf/compare/enable-postgres-extensions

Once this is merged I can raise the pull requests for the bosh-release, then paas-cf.

## Who

Anyone but me

henrytk committed:
Create Postgres extensions

We want users of Postgres instances to be able to use popular extensions[1].

Given there is little or no risk in enabling unneeded extensions we can
simplify the experience for users by doing it for them.

The options we explored:

* enabling them at the AWS level. This is not possible, as database parameter
groups do not allow you to configure extensions in any way.
* enabling them on `bind` operations. This ensures the instance is ready to
have these SQL statements run, but conceptually does not fit, as it is nothing
to do with the binding process.
* using the last operation polling: by waiting for the instance to be ready
and for all post-restore tasks (from snapshots), to be completed we ensure the
instance is in a state to have the necessary SQL statements run. It also ensures
the instance isn't used by users until the extensions are created.

We went with the latter option.

[1] https://www.pivotaltracker.com/n/projects/1275640/stories/148570883

henrytk committed:
Remove unnecessary variable declaration

This is only required for the outermost `Describe` call.

henrytk committed:
Use pointer for passing around DB details

This method makes no alterations to the database details, so we can
pass a pointer and operate on the actual data structure, instead of
copying a hefty struct all over the place.

henrytk commented on sqlengine/postgres_engine.go:
```diff
@@ -317,6 +317,22 @@ func (d *PostgresEngine) JDBCURI(address string, port int64, dbname string, user
+
+func (d *PostgresEngine) CreateExtensions(extensions []string) error {
+	for _, extension := range extensions {
+		createExtensionTemplate := template.Must(template.New(extension + "Extension").Parse(createExtensionPattern))
```
To gather the reviewer's opinion: I used a template here for consistency with the other SQL statements, but given the simplicity of this statement I am wondering if the standard SQL placeholders in the `database/sql` package would have been more appropriate.

46bit replied:
I felt unsure so tried writing out the code.
```go
func (d *PostgresEngine) CreateExtensions(extensions []string) error {
	for _, extension := range extensions {
		_, err := d.db.Exec("CREATE EXTENSION IF NOT EXISTS $1", extension)
		if err != nil {
			return err
		}
	}
	return nil
}
```
This errors with something implying Postgres isn't recognising the query parameter:
```
  Expected error:
      <*pq.Error | 0xc420493b00>: {
          Severity: "ERROR",
          Code: "42601",
          Message: "syntax error at or near \"'$1'\"",
          Detail: "",
          Hint: "",
          Position: "32",
          InternalPosition: "",
          InternalQuery: "",
          Where: "",
          Schema: "",
          Table: "",
          Column: "",
          DataTypeName: "",
          Constraint: "",
          File: "scan.l",
          Line: "1082",
          Routine: "scanner_yyerror",
      }
      pq: syntax error at or near "'$1'"
  not to have occurred
```

After some time I haven't reached a firm conclusion whether the problem is me or Postgres. Unless you spot the issue, let's stick with the template?

henrytk replied:
The documentation is confusing, because [`QueryRow`](https://golang.org/pkg/database/sql/#DB.QueryRow) for example uses a `?` as a placeholder. Can't seem to find the relevant documentation on its proper usage.

46bit commented on ci/blackbox/config.json:
```diff
@@ -27,7 +27,11 @@
+								"postgres_extensions": [
+									"uuid-ossp",
+									"postgis"
+								]
```
This is using tabs where the rest of the file uses spaces - I'll push a fix.

46bit commented on ci/blackbox/rdsbroker_test.go:
```diff
@@ -387,3 +389,34 @@ func permissionsTest(databaseURI string) error {
+		}
+		Expect(rows.Err()).ToNot(HaveOccurred())
+		Expect(extensions).To(ContainElement("uuid-ossp"))
+		Expect(extensions).To(ContainElement("postgis"))
```
@henrytk It worries me a little that a typo could disable this test and we might never notice. Chris suggested adding a `case "mysql"` that does nothing and then a `default` case which fails, so that if a typo happened the tests would fail. What do you think?

henrytk replied:
Oooh, I like the default = fail approach. ~~Not sure we need a blank "mysql" case though.~~ <- I remember, we re-use the same for MySQL

46bit reviewed (commented):
By and large looks 👍 . I've highlighted something about the `postgresExtensionsTest` in the blackbox tests (see comment) but it's not critical.

henrytk committed:
Add extra guard in integration tests

This will guard against silent failure if the scheme is not 'postgres',
while allowing the same code to be used for MySQL tests.
